File: quasar-backend/src/controllers/skillController.js
const SkillVector = require('../models/SkillVector');
const Session = require('../models/Session');
const JdQuestion = require('../models/JdQuestion');
const JdSession = require('../models/JdSession');
const logger = require('../utils/logger');

// STAR dimension → skill mapping
const STAR_SKILL_MAP = {
  situation: 'communication',
  task: 'problem_structuring',
  action: 'leadership',
  result: 'result_orientation'
};

const CATEGORY_SKILL_MAP = {
  technical: 'technical_depth',
  'system-design': 'technical_depth',
  behavioural: 'communication',
  'culture-fit': 'culture_fit'
};

/**
 * POST /api/users/:userId/skill-vector/update
 * Update skill vector using exponential moving average
 */
async function updateSkillVector(req, res) {
  try {
    const { userId } = req.params;
    const { sessionId, starScores, categoryScores } = req.body;

    if (!starScores && !categoryScores) {
      return res.status(400).json({ success: false, message: 'starScores or categoryScores required', data: null });
    }

    const updates = [];

    // Map STAR scores to skills
    if (starScores) {
      for (const [dimension, score] of Object.entries(starScores)) {
        const skill = STAR_SKILL_MAP[dimension.toLowerCase()];
        if (skill && typeof score === 'number') {
          updates.push({ skill, score: Math.max(0, Math.min(10, score)) });
        }
      }
    }

    // Map category scores to skills
    if (categoryScores) {
      for (const [category, score] of Object.entries(categoryScores)) {
        const skill = CATEGORY_SKILL_MAP[category.toLowerCase()];
        if (skill && typeof score === 'number') {
          updates.push({ skill, score: Math.max(0, Math.min(10, score)) });
        }
      }
    }

    // Apply exponential moving average: newScore = 0.7 * oldScore + 0.3 * latestScore
    for (const { skill, score } of updates) {
      const existing = await SkillVector.findOne({ userId, skill });

      if (existing) {
        const oldScore = parseFloat(existing.score);
        const newScore = parseFloat((0.7 * oldScore + 0.3 * score).toFixed(2));
        const newCount = existing.attemptCount + 1;

        await SkillVector.findOneAndUpdate(
          { userId, skill },
          { score: newScore, attemptCount: newCount, lastUpdated: new Date() }
        );
      } else {
        await SkillVector.create({ userId, skill, score, attemptCount: 1 });
      }
    }

    // Return updated skill vector
    const result = await SkillVector.find({ userId }).sort({ skill: 1 }).lean();

    return res.json({
      success: true,
      message: 'Skill vector updated',
      data: result
    });
  } catch (err) {
    logger.error('Update skill vector error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to update skill vector', data: null });
  }
}

/**
 * GET /api/users/:userId/skill-vector
 * Get the user's current skill vector
 */
async function getSkillVector(req, res) {
  try {
    const { userId } = req.params;

    const result = await SkillVector.find({ userId }).sort({ skill: 1 }).lean();

    return res.json({
      success: true,
      message: 'Skill vector retrieved',
      data: result
    });
  } catch (err) {
    logger.error('Get skill vector error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get skill vector', data: null });
  }
}

/**
 * GET /api/sessions/:sessionId/next-question
 * Adaptive question selection based on skill vector + spaced repetition
 */
async function getNextQuestion(req, res) {
  try {
    const { sessionId } = req.params;

    // Get session to find user and JD
    const session = await Session.findById(sessionId).select('userId jdSessionId').lean();

    if (!session) {
      return res.status(404).json({ success: false, message: 'Session not found', data: null });
    }

    const userId = session.userId;
    const jdSessionId = session.jdSessionId;

    if (!userId) {
      return res.status(400).json({ success: false, message: 'Session has no user', data: null });
    }

    // Get user's skill vector
    const skills = await SkillVector.find({ userId }).sort({ score: 1 }).lean();

    // Find 2 lowest-scoring skills
    const weakestSkills = skills.slice(0, 2).map(s => s.skill);

    // Map skills back to question target_skill values
    const targetSkillTerms = new Set();
    for (const weakSkill of weakestSkills) {
      targetSkillTerms.add(weakSkill);
      // Also add reverse mappings
      for (const [dim, skill] of Object.entries(STAR_SKILL_MAP)) {
        if (skill === weakSkill) targetSkillTerms.add(dim);
      }
      for (const [cat, skill] of Object.entries(CATEGORY_SKILL_MAP)) {
        if (skill === weakSkill) targetSkillTerms.add(cat);
      }
    }

    // Build question query — prioritize from JD questions if available
    let questions;
    if (jdSessionId) {
      questions = await JdQuestion.find({ jdSessionId }).sort({ weight: -1 }).lean();
    } else {
      // Default: generic questions from any JD for this user
      const jdSessions = await JdSession.find({ userId }).select('_id').lean();
      const jdIds = jdSessions.map(j => j._id);
      questions = jdIds.length > 0
        ? await JdQuestion.find({ jdSessionId: { $in: jdIds } }).sort({ weight: -1 }).limit(50).lean()
        : [];
    }

    if (questions.length === 0) {
      return res.json({
        success: true,
        message: 'No questions available',
        data: null
      });
    }

    // Count completed sessions for spaced repetition
    const totalSessions = await Session.countDocuments({ userId, status: 'completed' });

    // Score each question
    const scored = questions.map(q => {
      let priority = q.weight || 0.5;

      // Boost if targeting weak skill
      const targetLower = (q.targetSkill || '').toLowerCase();
      if (targetSkillTerms.has(targetLower)) {
        priority += 0.3;
      }

      // Spaced repetition boost: not attempted recently + low score
      if (!q.lastAttemptedSession) {
        priority += 0.2; // Never attempted
      } else if (q.lastScore !== null && parseFloat(q.lastScore) < 6) {
        priority += 0.15;
      }

      return { ...q, priority };
    });

    // Sort by priority desc, pick top
    scored.sort((a, b) => b.priority - a.priority);
    const selected = scored[0];

    // Determine reason
    let reason = 'Highest priority question';
    const targetLower = (selected.targetSkill || '').toLowerCase();
    if (targetSkillTerms.has(targetLower)) {
      const matchedWeak = weakestSkills.find(ws => {
        if (ws === targetLower) return true;
        for (const [k, v] of Object.entries({ ...STAR_SKILL_MAP, ...CATEGORY_SKILL_MAP })) {
          if (v === ws && k === targetLower) return true;
        }
        return false;
      });
      if (matchedWeak) {
        const weakScore = skills.find(s => s.skill === matchedWeak)?.score || 5;
        reason = `Targets your weakest skill: ${matchedWeak} (score: ${weakScore}/10)`;
      }
    }

    return res.json({
      success: true,
      message: 'Next question selected',
      data: {
        question: selected.question,
        difficulty: selected.difficulty,
        targetSkill: selected.targetSkill,
        category: selected.category,
        reason,
        questionId: selected._id
      }
    });
  } catch (err) {
    logger.error('Get next question error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get next question', data: null });
  }
}

module.exports = { updateSkillVector, getSkillVector, getNextQuestion };

File: backend/src/main/java/com/teamzemo/quasar/controller/SkillController.java
package com.teamzemo.quasar.controller;

import com.teamzemo.quasar.dto.ApiResponse;
import com.teamzemo.quasar.dto.UpdateSkillVectorRequest;
import com.teamzemo.quasar.model.JdQuestion;
import com.teamzemo.quasar.model.JdSession;
import com.teamzemo.quasar.model.Session;
import com.teamzemo.quasar.model.SkillVector;
import com.teamzemo.quasar.repository.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api")
public class SkillController {

    private static final Logger log = LoggerFactory.getLogger(SkillController.class);

    private final SkillVectorRepository  skillVectorRepository;
    private final SessionRepository       sessionRepository;
    private final JdQuestionRepository    jdQuestionRepository;
    private final JdSessionRepository     jdSessionRepository;

    // STAR dimension → skill mapping (mirrors JS STAR_SKILL_MAP)
    private static final Map<String, String> STAR_SKILL_MAP = Map.of(
            "situation", "communication",
            "task",      "problem_structuring",
            "action",    "leadership",
            "result",    "result_orientation"
    );

    // Category → skill mapping (mirrors JS CATEGORY_SKILL_MAP)
    private static final Map<String, String> CATEGORY_SKILL_MAP = Map.of(
            "technical",    "technical_depth",
            "system-design","technical_depth",
            "behavioural",  "communication",
            "culture-fit",  "culture_fit"
    );

    public SkillController(SkillVectorRepository skillVectorRepository,
                            SessionRepository sessionRepository,
                            JdQuestionRepository jdQuestionRepository,
                            JdSessionRepository jdSessionRepository) {
        this.skillVectorRepository = skillVectorRepository;
        this.sessionRepository     = sessionRepository;
        this.jdQuestionRepository  = jdQuestionRepository;
        this.jdSessionRepository   = jdSessionRepository;
    }

    /** POST /api/users/:userId/skill-vector/update */
    @PostMapping("/users/{userId}/skill-vector/update")
    public ResponseEntity<ApiResponse<List<SkillVector>>> updateSkillVector(
            @PathVariable String userId,
            @RequestBody UpdateSkillVectorRequest req) {
        try {
            if (req.getStarScores() == null && req.getCategoryScores() == null) {
                return ResponseEntity.badRequest()
                        .body(ApiResponse.fail("starScores or categoryScores required"));
            }

            List<Map.Entry<String, Double>> updates = new ArrayList<>();

            if (req.getStarScores() != null) {
                for (Map.Entry<String, Object> e : req.getStarScores().entrySet()) {
                    String skill = STAR_SKILL_MAP.get(e.getKey().toLowerCase());
                    if (skill != null && e.getValue() instanceof Number n) {
                        updates.add(Map.entry(skill, clamp(n.doubleValue())));
                    }
                }
            }

            if (req.getCategoryScores() != null) {
                for (Map.Entry<String, Object> e : req.getCategoryScores().entrySet()) {
                    String skill = CATEGORY_SKILL_MAP.get(e.getKey().toLowerCase());
                    if (skill != null && e.getValue() instanceof Number n) {
                        updates.add(Map.entry(skill, clamp(n.doubleValue())));
                    }
                }
            }

            for (Map.Entry<String, Double> update : updates) {
                applyEMA(userId, update.getKey(), update.getValue());
            }

            List<SkillVector> result = skillVectorRepository.findByUserIdOrderBySkillAsc(userId);
            return ResponseEntity.ok(ApiResponse.ok("Skill vector updated", result));
        } catch (Exception e) {
            log.error("Update skill vector error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to update skill vector"));
        }
    }

    /** GET /api/users/:userId/skill-vector */
    @GetMapping("/users/{userId}/skill-vector")
    public ResponseEntity<ApiResponse<List<SkillVector>>> getSkillVector(@PathVariable String userId) {
        try {
            List<SkillVector> result = skillVectorRepository.findByUserIdOrderBySkillAsc(userId);
            return ResponseEntity.ok(ApiResponse.ok("Skill vector retrieved", result));
        } catch (Exception e) {
            log.error("Get skill vector error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get skill vector"));
        }
    }

    /** GET /api/sessions/:sessionId/next-question */
    @GetMapping("/sessions/{sessionId}/next-question")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getNextQuestion(@PathVariable String sessionId) {
        try {
            Session session = sessionRepository.findById(sessionId).orElse(null);
            if (session == null) {
                return ResponseEntity.status(404).body(ApiResponse.fail("Session not found"));
            }

            String userId      = session.getUserId();
            String jdSessionId = session.getJdSessionId();

            if (userId == null) {
                return ResponseEntity.badRequest().body(ApiResponse.fail("Session has no user"));
            }

            // Get weakest skills
            List<SkillVector> skills = skillVectorRepository.findByUserIdOrderByScoreAsc(userId);
            List<String> weakestSkills = skills.stream().limit(2)
                    .map(SkillVector::getSkill).collect(Collectors.toList());

            // Build target skill terms
            Set<String> targetSkillTerms = new HashSet<>(weakestSkills);
            for (String ws : weakestSkills) {
                STAR_SKILL_MAP.forEach((k, v) -> { if (v.equals(ws)) targetSkillTerms.add(k); });
                CATEGORY_SKILL_MAP.forEach((k, v) -> { if (v.equals(ws)) targetSkillTerms.add(k); });
            }

            // Load questions
            List<JdQuestion> questions;
            if (jdSessionId != null) {
                questions = jdQuestionRepository.findByJdSessionIdOrderByWeightDesc(jdSessionId);
            } else {
                List<String> jdIds = jdSessionRepository.findByUserId(userId).stream()
                        .map(JdSession::getId).collect(Collectors.toList());
                questions = jdIds.isEmpty() ? List.of()
                        : jdQuestionRepository.findByJdSessionIdInOrderByWeightDesc(jdIds)
                                .stream().limit(50).collect(Collectors.toList());
            }

            if (questions.isEmpty()) {
                return ResponseEntity.ok(ApiResponse.ok("No questions available", null));
            }

            // Score each question
            List<Map.Entry<JdQuestion, Double>> scored = questions.stream().map(q -> {
                double priority = q.getWeight() != null ? q.getWeight() : 0.5;
                String tl = q.getTargetSkill() != null ? q.getTargetSkill().toLowerCase() : "";
                if (targetSkillTerms.contains(tl)) priority += 0.3;
                if (q.getLastAttemptedSession() == null)   priority += 0.2;
                else if (q.getLastScore() != null && q.getLastScore() < 6) priority += 0.15;
                return Map.entry(q, priority);
            }).sorted((a, b) -> Double.compare(b.getValue(), a.getValue()))
              .collect(Collectors.toList());

            JdQuestion selected  = scored.get(0).getKey();
            String     targetLow = selected.getTargetSkill() != null
                    ? selected.getTargetSkill().toLowerCase() : "";

            String reason = "Highest priority question";
            if (targetSkillTerms.contains(targetLow)) {
                String matchedWeak = weakestSkills.stream()
                        .filter(ws -> ws.equals(targetLow)
                                || STAR_SKILL_MAP.getOrDefault(targetLow, "").equals(ws)
                                || CATEGORY_SKILL_MAP.getOrDefault(targetLow, "").equals(ws))
                        .findFirst().orElse(null);
                if (matchedWeak != null) {
                    double weakScore = skills.stream()
                            .filter(sv -> sv.getSkill().equals(matchedWeak))
                            .mapToDouble(SkillVector::getScore).findFirst().orElse(5.0);
                    reason = "Targets your weakest skill: " + matchedWeak + " (score: " + weakScore + "/10)";
                }
            }

            return ResponseEntity.ok(ApiResponse.ok("Next question selected", Map.of(
                    "question",    selected.getQuestion(),
                    "difficulty",  selected.getDifficulty(),
                    "targetSkill", selected.getTargetSkill() != null ? selected.getTargetSkill() : "",
                    "category",    selected.getCategory() != null ? selected.getCategory() : "",
                    "reason",      reason,
                    "questionId",  selected.getId()
            )));
        } catch (Exception e) {
            log.error("Get next question error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get next question"));
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────

    private void applyEMA(String userId, String skill, double score) {
        SkillVector existing = skillVectorRepository.findByUserIdAndSkill(userId, skill).orElse(null);
        if (existing != null) {
            double newScore = Math.round((0.7 * existing.getScore() + 0.3 * score) * 100.0) / 100.0;
            existing.setScore(newScore);
            existing.setAttemptCount(existing.getAttemptCount() + 1);
            existing.setLastUpdated(Instant.now());
            skillVectorRepository.save(existing);
        } else {
            SkillVector sv = new SkillVector();
            sv.setUserId(userId);
            sv.setSkill(skill);
            sv.setScore(score);
            sv.setAttemptCount(1);
            skillVectorRepository.save(sv);
        }
    }

    private double clamp(double v) { return Math.max(0, Math.min(10, v)); }
}
